Comparing int32 value with int causes ambiguous call error

testints.nim:
```nim
var x: int32
var y: int

# Following code compiles without errors.
discard x == y
discard y == 123.int32
discard 123.int32 == y
discard 123.int == 123.int
discard 123.int32 == 123.int32

# Following code causes compile error: "Error: ambiguous call"
discard 123.int32 == 123
discard 123 == 123.int32
discard x == 123
```

```console
$ ./bin/nimony c mytests/testints.nim 
nif/bin/nimsem  --isSystem m nifcache/sys9azlf.1.nif nifcache/sys9azlf.2.nif nifcache/sys9azlf.2.idx.nif
nif/bin/nimsem  m nifcache/tes9vq2bh.1.nif nifcache/tes9vq2bh.2.nif nifcache/tes9vq2bh.2.idx.nif
Traceback (most recent call last)
nif/src/nimony/nimsem.nim(120) nimsem
nif/src/nimony/nimsem.nim(117) handleCmdLine
nif/src/nimony/nimsem.nim(50) singleModule
nif/src/nimony/sem.nim(5443) semcheck
nif/src/nimony/sem.nim(746) semStmt
nif/src/nimony/sem.nim(5159) semExpr
nif/src/nimony/sem.nim(3455) semDiscard
nif/src/nimony/sem.nim(5233) semExpr
nif/src/nimony/sem.nim(1609) semCall
nif/src/nimony/sem.nim(1441) resolveOverloads
nif/src/nimony/sembasics.nim(159) buildErr
testints.nim(12, 19) Error: ambiguous call
ambiguous call
make: *** [nifcache/tes9vq2bh.makefile:6: nifcache/tes9vq2bh.2.idx.nif] Error 1
FAILURE: make -f nifcache/tes9vq2bh.makefile
```

--- a/src/nimony/sigmatch.nim
+++ b/src/nimony/sigmatch.nim
@@ -56,7 +56,7 @@ type
     skippedMod: TypeKind
     argInfo: PackedLineInfo
     pos, opened: int
-    inheritanceCosts, intCosts: int
+    inheritanceCosts, intLitCosts, intConvCosts, convCosts: int
     returnType*: Cursor
     context: ptr SemContext
     error: MatchError
@@ -513,7 +513,14 @@ proc matchIntegralType(m: var Match; f: var Cursor; arg: Item) =
     else:
       m.args.addParLe HconvX, m.argInfo
       m.args.addSubtree forig
-      inc m.intCosts
+      if isIntLit:
+        if f.typeKind == FloatT:
+          inc m.convCosts
+        else:
+          inc m.intLitCosts
+      else:
+        # sameKind is true
+        inc m.intConvCosts
       inc m.opened
   else:
     m.error InvalidMatch, f, a
@@ -618,7 +625,7 @@ proc singleArgImpl(m: var Match; f: var Cursor; arg: Item) =
         m.args.addParLe HconvX, m.argInfo
         m.args.addSubtree f
         inc m.opened
-        inc m.intCosts
+        inc m.convCosts
         inc f
         expectParRi m, f
       else:
@@ -634,7 +641,7 @@ proc singleArgImpl(m: var Match; f: var Cursor; arg: Item) =
         m.args.addParLe HconvX, m.argInfo
         m.args.addSubtree f
         inc m.opened
-        inc m.intCosts
+        inc m.convCosts
         inc f
         expectParRi m, f
       else:
@@ -715,19 +722,25 @@ proc typematch*(m: var Match; formal: Cursor; arg: Item) =
 type
   TypeRelation* = enum
     NoMatch
+    IntLitMatch
+    IntConvMatch
     ConvertibleMatch
     SubtypeMatch
     GenericMatch
     EqualMatch
 
 proc usesConversion*(m: Match): bool {.inline.} =
-  result = abs(m.inheritanceCosts) + m.intCosts > 0
+  result = abs(m.inheritanceCosts) + m.intLitCosts + m.intConvCosts + m.convCosts > 0
 
 proc classifyMatch*(m: Match): TypeRelation {.inline.} =
   if m.err:
     return NoMatch
-  if m.intCosts != 0:
+  if m.convCosts != 0:
     return ConvertibleMatch
+  if m.intConvCosts != 0:
+    return IntConvMatch
+  if m.intLitCosts != 0:
+    return IntLitMatch
   if m.inheritanceCosts != 0:
     return SubtypeMatch
   if m.inferred.len != 0:
@@ -864,13 +877,21 @@ type
 proc cmpMatches*(a, b: Match): DisambiguationResult =
   assert not a.err
   assert not b.err
-  if a.inheritanceCosts < b.inheritanceCosts:
+  if a.convCosts < b.convCosts:
+    result = FirstWins
+  elif a.convCosts > b.convCosts:
+    result = SecondWins
+  elif a.intConvCosts < b.intConvCosts:
+    result = FirstWins
+  elif a.intConvCosts > b.intConvCosts:
+    result = SecondWins
+  elif a.intLitCosts < b.intLitCosts:
+    result = FirstWins
+  elif a.intLitCosts > b.intLitCosts:
+    result = SecondWins
+  elif a.inheritanceCosts < b.inheritanceCosts:
     result = FirstWins
   elif a.inheritanceCosts > b.inheritanceCosts:
-    result = SecondWins
-  elif a.intCosts < b.intCosts:
-    result = FirstWins
-  elif a.intCosts > b.intCosts:
     result = SecondWins
   else:
     let diff = a.inferred.len - b.inferred.len

--- a/tests/nimony/sysbasics/tintlitmatch.nif
+++ b/tests/nimony/sysbasics/tintlitmatch.nif
@@ -1,0 +1,89 @@
+(.nif24)
+0,1,tests/nimony/sysbasics/tintlitmatch.nim(stmts
+ (proc 5 :foo.0.tin4pj0od1 . . . 8
+  (params 1
+   (param :x.0 . . 3
+    (i +16) .)) . . . 21
+  (stmts
+   (discard .))) ,1
+ (proc 5 :bar.0.tin4pj0od1 . . . 8
+  (params 1
+   (param :x.1 . . 3
+    (f +64) .)) . . . 21
+  (stmts
+   (discard .))) 3,2
+ (call ~3 foo.0.tin4pj0od1 1
+  (hconv 8,~2
+   (i +16) +123)) 3,3
+ (call ~3 bar.0.tin4pj0od1 1
+  (hconv 8,~2
+   (f +64) +123)) ,4
+ (proc 5 :baz.0.tin4pj0od1 . . . 8
+  (params 1
+   (param :x.2 . . 3
+    (i +16) .)) . . . 21
+  (stmts
+   (discard .))) ,5
+ (proc 5 :baz.1.tin4pj0od1 . . . 8
+  (params 1
+   (param :x.3 . . 3
+    (i -1) .)) . . . 19
+  (stmts
+   (discard .))) 3,6
+ (call ~3 baz.1.tin4pj0od1 1 +123) 4,9
+ (var :x.0.tin4pj0od1 . . 3
+  (i +32) .) 4,10
+ (var :y.0.tin4pj0od1 . . 3
+  (i -1) .) ,13
+ (discard 10
+  (eq ~3,~4
+   (i +32) ~2
+   (hconv 22,685,lib/std/system.nim
+    (i +64) x.0.tin4pj0od1) 3 y.0.tin4pj0od1)) ,14
+ (discard 10
+  (eq ~3,~4
+   (i -1) ~2 y.0.tin4pj0od1 6
+   (hconv 22,685,lib/std/system.nim
+    (i +64)
+    (conv 1
+     (i +32) ~3 +123)))) ,15
+ (discard 18
+  (eq ~1,~1
+   (i +32) ~7
+   (hconv 22,685,lib/std/system.nim
+    (i +64)
+    (conv 6,~1
+     (i +32) ~3 +123)) 3 y.0.tin4pj0od1)) ,16
+ (discard 16
+  (eq ~4
+   (i -1) ~5
+   (conv 1
+    (i -1) ~3 +123) 6
+   (conv ~10
+    (i -1) ~3 +123))) ,17
+ (discard 18
+  (eq ~1,~3
+   (i +32) ~7
+   (conv 6,~3
+    (i +32) ~3 +123) 6
+   (conv ~7,~3
+    (i +32) ~3 +123))) ,20
+ (discard 18
+  (eq ~1,~6
+   (i +32) ~7
+   (conv 6,~6
+    (i +32) ~3 +123) 3
+   (hconv 22,683,lib/std/system.nim
+    (i +32) +123))) ,21
+ (discard 12
+  (eq
+   (i -1) ~4
+   (hconv 22,683,lib/std/system.nim
+    (i +32) +123) 6
+   (conv ~1,~7
+    (i +32) ~3 +123))) ,22
+ (discard 10
+  (eq ~3,~13
+   (i +32) ~2 x.0.tin4pj0od1 3
+   (hconv 22,683,lib/std/system.nim
+    (i +32) +123))))

--- a/tests/nimony/sysbasics/tintlitmatch.nim
+++ b/tests/nimony/sysbasics/tintlitmatch.nim
@@ -5,3 +5,19 @@ bar(123)
 proc baz(x: int16) = discard
 proc baz(x: int) = discard
 baz(123)
+
+# issue #477
+var x: int32
+var y: int
+
+# Following code compiles without errors.
+discard x == y
+discard y == 123.int32
+discard 123.int32 == y
+discard 123.int == 123.int
+discard 123.int32 == 123.int32
+
+# Following code causes compile error: "Error: ambiguous call"
+discard 123.int32 == 123
+discard 123 == 123.int32
+discard x == 123

